File: RemindMePlease-Swift/Sources/RemindMePlease/Storage.swift
import Foundation

final class Storage {
    static let shared = Storage()

    private let dataDir: URL
    private let dataFile: URL

    private let defaultData: [String: Any] = [
        "tasks": [Any](),
        "categories": ["Work", "Personal", "Urgent", "Design", "Marketing"],
        "weeklyHistory": [Any](),
        "settings": [
            "keyboardShortcut": "CommandOrControl+Shift+Space",
            "theme": "dark",
            "defaultView": "master",
            "useIcons": false
        ] as [String: Any],
        "quickNote": ""
    ]

    private init() {
        dataDir  = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent(".remindmeplease")
        dataFile = dataDir.appendingPathComponent("tasks.json")
    }

    var folderPath: String { dataDir.path }

    func read() -> [String: Any] {
        ensureDir()
        guard
            FileManager.default.fileExists(atPath: dataFile.path),
            let raw  = try? Data(contentsOf: dataFile),
            let json = try? JSONSerialization.jsonObject(with: raw) as? [String: Any]
        else {
            write(defaultData)
            return defaultData
        }
        return json
    }

    @discardableResult
    func write(_ data: [String: Any]) -> Bool {
        ensureDir()
        guard let encoded = try? JSONSerialization.data(
            withJSONObject: data, options: .prettyPrinted
        ) else { return false }
        try? encoded.write(to: dataFile)
        return true
    }

    private func ensureDir() {
        try? FileManager.default.createDirectory(
            at: dataDir, withIntermediateDirectories: true)
    }
}

File: src/storage.js
const fs = require('fs')
const path = require('path')
const os = require('os')

const DEV = process.env.RMP_DEV === 'true'
const DATA_DIR = DEV
  ? path.join(__dirname, '..', 'data')
  : path.join(os.homedir(), '.remindmeplease')
const DATA_FILE = path.join(DATA_DIR, 'tasks.json')

const DEFAULT_DATA = {
  tasks: [],
  categories: ['Work', 'Personal', 'Urgent', 'Design', 'Marketing'],
  weeklyHistory: [],
  settings: {
    keyboardShortcut: 'CommandOrControl+Shift+Space',
    theme: 'dark',
    defaultView: 'list',
    listViewMode: 'list',
    useIcons: false,
    pomodoro: {
      workMinutes: 25,
      shortBreakMinutes: 5,
      longBreakMinutes: 20,
      sessionsBeforeLongBreak: 4
    }
  },
  quickNote: ''
}

function ensureDir () {
  if (!fs.existsSync(DATA_DIR)) fs.mkdirSync(DATA_DIR, { recursive: true })
}

function read () {
  ensureDir()
  if (!fs.existsSync(DATA_FILE)) {
    fs.writeFileSync(DATA_FILE, JSON.stringify(DEFAULT_DATA, null, 2))
    return DEFAULT_DATA
  }
  try {
    return JSON.parse(fs.readFileSync(DATA_FILE, 'utf8'))
  } catch {
    return DEFAULT_DATA
  }
}

function write (data) {
  ensureDir()
  fs.writeFileSync(DATA_FILE, JSON.stringify(data, null, 2))
  return true
}

function getDataFilePath () {
  return DATA_FILE
}

module.exports = { read, write, getDataFilePath }
